bindings/pyroot_experimental/pythonizations/python/ROOT/_numbadeclare.py

```diff
@@ -0,0 +1,278 @@
+
+        # Define function call
+        vecbool_conversion = []
+        func_args = []
+        for i, t in enumerate(input_types):
+            if 'RVec' in t:
+                func_args += ['x_{0}.data(), x_{0}.size()'.format(i)]
+                if get_inner_type(t) == 'bool':
+                    # Copy the RVec<bool> to a RVec<char> to match the numpy memory layout
+                    func_args[-1] = func_args[-1].replace('x_', 'xb_')
+                    vecbool_conversion += ['ROOT::RVec<char> xb_{0} = x_{0};'.format(i)]
+            else:
+                func_args += ['x_{}'.format(i)]
+        if 'RVec' in return_type:
+            # See C++ wrapper code for the reason using these arguments
+            func_args += ['&ptr, &size, dtype_size']
+
+        # Define return operation
+        if 'RVec' in return_type:
+            innert = get_inner_type(return_type)
+            if innert == 'bool': innert = 'char'
+            return_op = '\n    '.join([
+                '// Because an RVec cannot take the ownership of external data, we have to copy the returned array',
+                'long size; // Size of the returned array',
+                '{}* ptr; // Pointer to the data of the returned array'.format(innert),
+                'int dtype_size = sizeof({}); // Size of array datatype in bytes'.format(innert),
+                'funcptr({});'.format(', '.join(func_args)),
+                'ROOT::RVec<{}> x_r(size);'.format(innert),
+                'memcpy(x_r.data(), ptr, size * dtype_size);',
+                'free(ptr);',
+                # If we return a RVec<bool>, we rely here on the automatic conversion of RVec<char> to RVec<bool>
+                'return x_r;'])
+        else:
+            return_op = 'return funcptr({});'.format(', '.join(func_args))
+
+        # Build wrapper code
+        cppwrappercode = """\
+#pragma cling optimize(3)
+namespace Numba {{
+/*
+ * C++ wrapper function around the jitted Python wrapping which calls the jitted Python callable
+ */
+{RETURN_TYPE} {FUNC_NAME}({INPUT_SIGNATURE}) {{
+    // Create a function pointer from the jitted Python wrapper
+    const auto funcptr = reinterpret_cast<{FUNC_PTR_TYPE}>({FUNC_PTR});
+    // Perform conversion of RVec<bool>
+    {VECBOOL_CONVERSION}
+    // Return the result
+    {RETURN_OP}
+}}
+}}""".format(
+                RETURN_TYPE='ROOT::' + return_type if 'RVec' in return_type else return_type,
+                FUNC_NAME=name,
+                INPUT_SIGNATURE=input_signature,
+                FUNC_PTR=address,
+                FUNC_PTR_TYPE=func_ptr_type,
+                VECBOOL_CONVERSION='\n    '.join(vecbool_conversion),
+                RETURN_OP=return_op)
+
+        # Jit wrapper C++ code
+        err = gbl_namespace.gInterpreter.Declare(cppwrappercode)
+        if not err:
+            raise Exception('Failed to jit C++ wrapper code with cling:\n{}'.format(cppwrappercode))
+        func.__cpp_wrapper__ = cppwrappercode
+
+        return func
```

Comment:
should we return the numba-jitted function instead? Is it less surprising if `@Numba.Declare` returns a numba-jitted function or the function itself?

Reply:
Hmm, ... I don't know!

Reply:
i think either solution is fine as long as the documentation of `ROOT.Numba.Declare` says what the decorator does to the function (either nothing, or it transforms it into a numba function).

Which makes me think: are there any user docs for this feature?

Reply:
There's a tutorial :D And I'll add the numba jitted function as an attribute!

Reply:
Now you can get from the callable `f` the numba jitted func as `f.jitted`. Any better suggestions for the wording?

Reply:
Probably, `f.jitted_func`?

Reply:
> There's a tutorial

There is a discoverability issue (users don't go look for a tutorial for a feature they don't know), but it's not for this PR to solve it. Maybe we should have a "Using RDF from python" section in the RDF user guide that points to things like `Numba.Declare`.

`f.jitted[_func]` could refer to either numba's jitted function or cling's. I'd suggest  something like `f.numba_func`

Reply:
Taken!

Reply:
Taken the `f.numba_func` and also added a rather massiv docstring to the outer decorator.

Reply:
great!